rasterio/__init__.py

```diff
@@ -143,42 +144,11 @@ def open(
     an Apache Commons VFS style string beginning with "zip://" or
     "tar://"". In this case, the ``path`` must be an absolute path
     within that container.
-
     """
-    if not isinstance(path, string_types):
-        raise TypeError("invalid path: %r" % path)
-    if mode and not isinstance(mode, string_types):
-        raise TypeError("invalid mode: %r" % mode)
-    if driver and not isinstance(driver, string_types):
-        raise TypeError("invalid driver: %r" % driver)
-
-    if transform:
-        transform = guard_transform(transform)
-    elif 'affine' in kwargs:
-        affine = kwargs.pop('affine')
-        transform = guard_transform(affine)
-
-    if mode == 'r':
-        from rasterio._io import RasterReader
-        s = RasterReader(path)
-    elif mode == 'r+':
-        from rasterio._io import writer
-        s = writer(path, mode)
-    elif mode == 'r-':
-        from rasterio._base import DatasetReader
-        s = DatasetReader(path)
-    elif mode == 'w':
-        from rasterio._io import writer
-        s = writer(path, mode, driver=driver,
-                   width=width, height=height, count=count,
-                   crs=crs, transform=transform, dtype=dtype,
-                   nodata=nodata,
-                   **kwargs)
-    else:
-        raise ValueError(
-            "mode string must be one of 'r', 'r+', or 'w', not %s" % mode)
-    s.start()
-    return s
+    env = env or Env()
+    return env.open(path, mode=mode, driver=driver, width=width, height=height,
+                    count=count, crs=crs, transform=transform, dtype=dtype,
+                    nodata=nodata, env=env, **kwargs)
```

Comment:
Pass an `Env` to `rasterio.open()` if you want. Otherwise, an instance will be created by default and passed into `env.open()`.